platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractCollectionProperty.java

```diff
@@ -477,58 +468,128 @@ public String toString() {
-                return result.asType();
+            Value<Void> compositeResult = Value.present();
+            for (Collector<T> collector : getCollectors()) {
+                if (compositeResult.isMissing() && !isAbsentIgnoring(collector)) {
+                    // The property is missing so far and the argument is of add/addAll.
+                    // The property is going to be missing regardless of its value.
+                    continue;
+                }
+                Value<Void> result = collector.collectEntries(consumer, valueCollector, builder);
+                if (result.isMissing()) {
+                    // This is the argument of add/addAll and it is missing. It "poisons" the property (it becomes missing).
+                    // We discard all values and side effects gathered so far.
+                    builder = collectionFactory.get();
+                    compositeResult = result;
+                } else if (compositeResult.isMissing()) {
+                    assert isAbsentIgnoring(collector);
+                    // This is an argument of append/appendAll. It "recovers" the property from the "poisoned" state.
+                    // Entries are already in the builder.
+                    compositeResult = result;
+                } else {
+                    assert !compositeResult.isMissing();
+                    // Both the property so far and the current argument are present, just continue building the value.
+                    // Entries are already in the builder.
+                    compositeResult = compositeResult.withSideEffect(SideEffect.fixedFrom(result));
+                }
             }
-            return Value.of(Cast.<C>uncheckedNonnullCast(builder.build())).withSideEffect(SideEffect.fixedFrom(result));
-        }
-
-        @Override
-        public CollectionSupplier<T, C> plus(Collector<T> addedCollector) {
-            Collector<T> left = value.absentIgnoringIfNeeded(ignoreAbsent);
-            Collector<T> right = addedCollector;
-            PlusCollector<T> newCollector = new PlusCollector<>(left, right);
-            return new CollectingSupplier(newCollector);
+            if (compositeResult.isMissing()) {
+                return compositeResult.asType();
+            }
+            return Value.of(Cast.<C>uncheckedNonnullCast(builder.build())).withSideEffect(SideEffect.fixedFrom(compositeResult));
         }
 
         @Override
-        public CollectionSupplier<T, C> absentIgnoring() {
-            return ignoreAbsent ? this : new CollectingSupplier(value, true);
+        public CollectionSupplier<T, C> plus(Collector<T> addedCollector, boolean ignoreAbsent) {
+            Preconditions.checkState(collectors.size() == size, "Something has been appended to this collector already");
+            collectors.add(ignoreAbsentIfNeeded(addedCollector, ignoreAbsent));
+            return new CollectingSupplier<>(type, collectionFactory, valueCollector, collectors, size + 1);
```

Comment:
🤔 This is really confusing, we're adding to the shared `collectors` and also returning a new *value*? Is the point of the value to simply keep track of the invariant in line 574?

Reply:
> 🤔 This is really confusing, we're adding to the shared `collectors` and also returning a new _value_? Is the point of the value to simply keep track of the invariant in line 574?

No, we need the collector to be immutable for shallow copying and, perhaps, convention to work. So this collector is a fixed-length prefix of a shared buffer. This is to avoid copying the array every time and still be able to iterate over it in forward direction efficiently.

Reply:
Ah, I see. I think a more explicitly named factory method instead of the constructor would make the intent clearer here. Nice trick.